### DISCUSSION TITLE
Is there some method like bulk_insert_mappings which can insert two-dimension array data?

### DISCUSSION THREAD

**Discussion by flyly0755** [sqlalchemy/sqlalchemy#9105] [Usage Questions]
```python
results = [{'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 1}, 
           {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 2}, 
           {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 3},     
           {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 4},
           ...
]
Session = sessionmaker(bind=engine)
s = Session()
s.bulk_insert_mappings(Results,results)
```
Above is common usage of bulk_insert_mappings.
I want to know if thers is method that supports two-dimension array data insert such as 
```python
results = [[u'SEX', 0.0, u'M', 1], 
           [u'SEX', 0.0, u'M', 2], 
           [u'SEX', 0.0, u'M', 3],     
           [u'SEX', 0.0, u'M', 4],
           ...
]
```
which is less memory usage compare with dict element.
And i think this maybe more quick to insert to database.
some method with 3 parameters (table, keyvalue_map, two-dimension array data)?

**Comment by CaselIT** [comment]
Hi,

No, there is currently no way of inserting data in that format using sqlalchemy, you will need to create dict from it using something like `dict(zip(key_list, inner_list))`.
Depending on your db and driver in use, you may be able to write a sql insert string that use positional binding to use with `exec_driver_sql`, but the data may require to be a list of tuples instead of a list of list

**Comment by zzzeek** [comment]
> ```python
> results = [{'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 1}, 
>            {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 2}, 
>            {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 3},     
>            {'attribute': u'SEX', 'value_d': 0.0, 'value_s': u'M', 'sid': 4},
>            ...
> ]
> Session = sessionmaker(bind=engine)
> s = Session()
> s.bulk_insert_mappings(Results,results)
> ```
> 
> Above is common usage of bulk_insert_mappings. I want to know if thers is method that supports two-dimension array data insert such as
> 
> ```python
> results = [[u'SEX', 0.0, u'M', 1], 
>            [u'SEX', 0.0, u'M', 2], 
>            [u'SEX', 0.0, u'M', 3],     
>            [u'SEX', 0.0, u'M', 4],
>            ...
> ]
> ```
> 
> which is less memory usage compare with dict element.

for "memory" that's only a partial solution. if your system runs out of memory with 10M dictionaries, it will run out of memory with 100M arrays.    if you are going for huge amounts of data, you will have to batch the data from the datasource in any case no matter what.  Also, the internals of SQLAlchemy will be creating dictionaries for all of the data above in any case, so trying to save memory at the level of "i want to create array of arrays instead of dictionaries", step one would be use raw DBAPI cursor, dont use SQLAlchemy at all.   However, there's no reason for any of that, it's all premature (and IMO ineffective) optimization.

> And i think this maybe more quick to insert to database. some method with 3 parameters (table, keyvalue_map, two-dimension array data)?

if you want "quick", where you are attempting to see the difference between constructing Python arrays vs. dicts, use PostgreSQL [COPY](https://www.psycopg.org/psycopg3/docs/basic/copy.html).   Better yet, write your program in Rust (but still, use COPY.  anything else is massively slower, dicts/arrays are nothing).     Otherwise, this level of micro-profiling will miss the forest for the trees in a big way.

**Reply by zzzeek** [reply]
to give some perspective here, trying to micro-optimize between arrays and dicts is like, "I want to get there faster.  But I'm not going to take the highway, I will take the scenic route, but drive 1 MPH above the speed limit".  That's the scale of this level of solution (in Python).

**Reply by flyly0755** [reply]
thank you for your detailed reply:)
Now I use clickhouse which is Column-oriented storage, so I think data batch insert with array of arrays(every element is one column data, different with list i post before)
```python
[[u'SEX',u'SEX', u'SEX', u'SEX', ...], 
[1.0, 2.0, 3.0, 4.0, ...], 
[1, 2, 3, 4, ...],     
   ...
]
```
maybe a lot faster.
for example, with 1 million data record, every data with 10 columns, wheater just need loop 10 times to finish insert(one time with one column data of 1 million values)?
and with array of dicts, which need loop 1 million times(one time with one dict data of 10 key-values)?
I am not very familar with internal logic of batch insert:( so my understand as above maybe very wrong, welcome to point out my mistake^_^

**Reply by CaselIT** [reply]
the array as you wrote are still in row-major (meaning that the first axis is the row, the second is the column), so you would still insert one row at a time, iterating 1 milion times.

I don't know how clickhouse works, but it seems strange that you can insert data by columns, since how would it know that the 3rd element of the second column insert should be in the same row as the 3rd element in the first column insert?

In any case this is not sql standard and not supported by sqlalchemy. If a feature like this is supported by clickhouse you would ask the clickhouse sqlalchemy dialect deveper for support. Otherwise get the cursor from the sqlalchemy connection and you can run whatever operation you like that's supported by the clickhouse python driver: https://docs.sqlalchemy.org/en/14/faq/connections.html#faq-dbapi-connection

**Reply by zzzeek** [reply]
Overall you should be running real world profiles for things.   we have a great way to get into that which is the [performance](https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.performance) suite.  you can run [bulk_inserts.py](https://docs.sqlalchemy.org/en/14/_modules/examples/performance/bulk_inserts.html) and see a comparison of many different types of bulk insert, and also add your own, the runner will do time comparisons or Python function call count comparisons.  it can dump out a profile for each test so you can get a feel for the specific functions being used in each .    overall getting into timing and Python profiling would grant a lot more perspective on this issue.
